thunderbird/processes/wps_decompose_flow_vectors.py

```diff
@@ -0,0 +1,123 @@
+# Processor imports
+from pywps import (
+    Process,
+    LiteralInput,
+    ComplexInput,
+    LiteralOutput,
+    ComplexOutput,
+    FORMATS,
+    Format,
+)
+# Library imports
+from netCDF4 import Dataset
+import numpy as np
+import logging
+import os
+import sys
+
+# Tool imports
+from dp.decompose_flow_vectors import logger, decompose_flow_vectors
+from thunderbird.utils import is_opendap_url
+
+
+LOGGER = logging.getLogger("PYWPS")
+
+
+class DecomposeFlowVectors(Process):
+    def __init__(self):
+        inputs = [
+            ComplexInput(
+                "netcdf",
+                "Daily NetCDF Dataset",
+                abstract="NetCDF file",
+                min_occurs=1,
+                max_occurs=1,
+                supported_formats=[FORMATS.NETCDF, FORMATS.DODS],
+            ),
+            LiteralInput(
+                "variable",
+                "netCDF variable describing flow direction",
+                data_type="string",
+            ),
+            LiteralInput(
+                "dest_file",
+                "destination netCDF file",
+                data_type="string",
+            ),
+        ]
+        outputs = [
+            ComplexOutput(
+                "output",
+                "Output",
+                abstract="output netCDF file",
+                as_reference=True,
+                supported_formats=[FORMATS.NETCDF],
+            ),
+        ]
+
+        super(DecomposeFlowVectors, self).__init__(
+            self._handler,
+            identifier="decompose_flow_vectors",
+            title="Update NetCDF Metadata",
+            abstract="Process an indexed flow direction netCDF into a vectored netCDF suitable for ncWMS display",
+            store_supported=True,
+            status_supported=True,
+            inputs=inputs,
+            outputs=outputs,
+        )
+
+    def get_filepath(self, request):
+        path = request.inputs["netcdf"][0]
+        if is_opendap_url(path.url):
+            return path.url
+        elif path.file.endswith(".nc"):
+            return path.file
+        else:
+            raise ProcessError(
+                "You must provide a data source (opendap/netcdf). "
+                f"Inputs provided: {request.inputs}"
+            )
+
+    def _handler(self, request, response):
+        response.update_status("Starting Process", 0)
+
+        source_file = self.get_filepath(request)
+        variable = request.inputs["variable"][0].data
+        dest_file = os.path.join(self.workdir, request.inputs["dest_file"][0].data)
+
+        source = Dataset(source_file, "r", format="NETCDF4")
+
+        if not "lon" in source.dimensions or not "lat" in source.dimensions:
+            logger.debug(
+                "{} does not have latitude and longitude dimensions".format(
+                    source_file
+                )
+            )
+            source.close()
+            sys.exit()
+
+        if not variable in source.variables:
```

Comment:
While it's admirable that you're checking your input here, I can't help but notice that it's verbatim the same code that is in [the `decompose_flow_vector` module](https://github.com/pacificclimate/climate-explorer-data-prep/blob/master/scripts/decompose_flow_vectors#L18). Might you be able to simply call `main()` and reuse your existing error checking?

I'm also concerned about using `sys.exit()` to immediately close down the process. Is that how WPS Processes are supposed to behave (asking out of ignorance)? I would suspect that there is some reporting, clean up and... actually returning a response to the client that the WPS Process need to complete. Going straight to `sys.exit()` would circumvent that. But I'm not sure. @nikola-rados could you weigh in?

Reply:
@jameshiebert 
> Might you be able to simply call main() and reuse your existing error checking?

I have actually thought about this before and implemented it in [one of my previous commits](https://github.com/pacificclimate/thunderbird/blob/60eea947bbc2d8f6fd302e89375eced28b0f70fb/thunderbird/processes/wps_decompose_flow_vectors.py#L23), and it actually performed nice. However, I think we should still avoid calling `main(args)` between programs. 

In the commit, I had to create two additional objects([arguments](https://github.com/pacificclimate/thunderbird/blob/60eea947bbc2d8f6fd302e89375eced28b0f70fb/thunderbird/processes/wps_decompose_flow_vectors.py#L29) and [ArgumentParserMocker](https://github.com/pacificclimate/thunderbird/blob/60eea947bbc2d8f6fd302e89375eced28b0f70fb/thunderbird/processes/wps_decompose_flow_vectors.py#L35)) and a function([load_script](https://github.com/pacificclimate/thunderbird/blob/60eea947bbc2d8f6fd302e89375eced28b0f70fb/thunderbird/processes/wps_decompose_flow_vectors.py#L16)) just to call `main(args)`. This commit was made before refactoring `decompose_flow_vectors`, so I had to import the functionality right from `script` package. But since we have `decompose_flow_vectors` in `dp` package now, I think it is not necessary nor ideal to call `main()` from `script` package since importing a script without `py` extension cannot be operated by default. Additionally, `arguments` and `ArgumentParserMocker` just simply cheats to behave like `argparse`, which can be a temporary solution but could possibly lead to another problem if newer `ce-dataprep` version uses more `argparse` functionalities.

Reply:
We could consider reworking (and renaming) the main arg found in `ce-dataprep`.  That would allow us to avoid repeating the same setup code while ensure you don't have to create the `args` object yourself.

Reply:
Now that I have modified the input checking procedure in `wps_decompose_flow_vectors` by creating two functions([source_check](https://github.com/pacificclimate/thunderbird/blob/issue/12/thunderbird/processes/wps_decompose_flow_vectors.py#L79) and [variable_check](https://github.com/pacificclimate/thunderbird/blob/issue/12/thunderbird/processes/wps_decompose_flow_vectors.py#L115)), I do not think it is necessary to call `main(args)` since the procedure differs from what it looks like in `ce-dataprep`. But instead, I think I could include the functions in `ce-dataprep` and call them in `thunderbird`. However, my concern is the distinction between `sys` and `wps` for exception handling.  `sys.exit()` is now replaced to `ProcessError` in `wps_decompose_flow_vectors`, but I don't think we want this in `ce-dataprep`.

Reply:
I think moving the functions you created into `ce-dataprep` is a great idea!  Your concern over the usage of `sys.exit()` is valid but I think we can solve the issue.  I believe the intent behind the use of `sys.exit()` was to exit the script if the parameters were not compatible/valid.  Instead of this exit strategy, we could raise an exception.  Then, in `thunderbird` we could use a `try/except` block in anticipation of this exception.  If the case is triggered we could raise the error message from `ce-dataprep` as a `ProcessError` in `thunderbird`.

This is a just a thought.  I don't know if this is actually the best way to handle an exception coming from our package.  @jameshiebert do you have any thoughts on this?

Reply:
That sounds like a good way to handle it, to me. It will exit the script where it needs to (as an unhandled exception), and it  gives us the flexibility to use different exit strategies in `thunderbird`. :smile: